File: packages/core/src/micracode_core/model_catalog.py
"""Catalog of provider/model pairs the API will accept at runtime."""

from __future__ import annotations

from dataclasses import dataclass

import httpx

from .config import CoreConfig

ProviderId = str


@dataclass(frozen=True)
class _Model:
    id: str
    label: str
    family: str


@dataclass(frozen=True)
class _Provider:
    id: str
    label: str
    models: tuple[_Model, ...]


_PROVIDERS: tuple[_Provider, ...] = (
    _Provider(
        id="openai",
        label="OpenAI",
        models=(
            _Model(id="gpt-5.4", label="GPT-5.4", family="openai-chat"),
            _Model(id="gpt-5-mini", label="GPT-5 Mini", family="openai-chat"),
            _Model(id="gpt-4.1", label="GPT-4.1", family="openai-chat"),
        ),
    ),
    _Provider(
        id="gemini",
        label="Google Gemini",
        models=(
            _Model(id="gemini-2.5-flash", label="Gemini 2.5 Flash", family="gemini"),
            _Model(id="gemini-2.5-pro", label="Gemini 2.5 Pro", family="gemini"),
            _Model(id="gemini-2.5-flash-lite", label="Gemini 2.5 Flash Lite", family="gemini"),
        ),
    ),
)


def _provider(pid: str) -> _Provider | None:
    for p in _PROVIDERS:
        if p.id == pid:
            return p
    return None


def _has_model(provider: _Provider, model_id: str) -> bool:
    return any(m.id == model_id for m in provider.models)


def _provider_available(config: CoreConfig, pid: str) -> bool:
    if pid == "openai":
        return bool(config.openai_api_key)
    if pid == "gemini":
        return bool(config.google_api_key)
    return False


async def _fetch_ollama_models(base_url: str) -> list[str]:
    try:
        async with httpx.AsyncClient(timeout=3.0) as client:
            resp = await client.get(f"{base_url}/api/tags")
            resp.raise_for_status()
            data = resp.json()
            return [m["name"] for m in data.get("models", [])]
    except Exception:
        return []


async def list_catalog(config: CoreConfig) -> dict:
    """Serialise the registry for the public ``GET /v1/models`` endpoint."""
    providers = [
        {
            "id": p.id,
            "label": p.label,
            "available": _provider_available(config, p.id),
            "models": [{"id": m.id, "label": m.label} for m in p.models],
        }
        for p in _PROVIDERS
    ]

    ollama_models = await _fetch_ollama_models(config.ollama_base_url)
    if ollama_models:
        providers.append(
            {
                "id": "ollama",
                "label": "Ollama (local)",
                "available": True,
                "models": [{"id": name, "label": name} for name in ollama_models],
            }
        )

    default = _default_selection(config, ollama_models)

    return {
        "providers": providers,
        "default": {"provider": default[0], "model": default[1]},
    }


def _default_selection(
    config: CoreConfig, ollama_models: list[str] | None = None
) -> tuple[str, str]:
    env_provider = config.llm_provider
    env_model = config.active_model

    if env_provider == "ollama":
        if env_model:
            return ("ollama", env_model)
        if ollama_models:
            return ("ollama", ollama_models[0])
    else:
        env = _provider(env_provider)
        if env is not None and env_model and _has_model(env, env_model):
            return (env_provider, env_model)

    for p in _PROVIDERS:
        if _provider_available(config, p.id) and p.models:
            return (p.id, p.models[0].id)

    if ollama_models:
        return ("ollama", ollama_models[0])

    first = _PROVIDERS[0]
    return (first.id, first.models[0].id)


def _model_family(provider: str, model_id: str) -> str:
    """Return the family string for a validated (provider, model_id) pair."""
    if provider == "ollama":
        return "ollama"
    p = _provider(provider)
    if p is not None:
        for m in p.models:
            if m.id == model_id:
                return m.family
    return "openai-chat"


def resolve(
    provider: str | None,
    model: str | None,
    config: CoreConfig,
) -> tuple[str, str, str]:
    """Validate a requested ``(provider, model)`` pair, filling in defaults.

    Returns a ``(provider, model, family)`` triple.
    """
    if provider is None and model is None:
        prov, mdl = _default_selection(config)
        return (prov, mdl, _model_family(prov, mdl))

    if provider is None or model is None:
        raise ValueError(
            "Both 'provider' and 'model' must be supplied together; got "
            f"provider={provider!r} model={model!r}."
        )

    if provider == "ollama":
        if not model:
            raise ValueError("model must be non-empty for provider 'ollama'.")
        return ("ollama", model, "ollama")

    p = _provider(provider)
    if p is None:
        known = ", ".join(pp.id for pp in _PROVIDERS) + ", ollama"
        raise ValueError(
            f"Unknown provider {provider!r}; supported providers: {known}."
        )

    if not _has_model(p, model):
        allowed = ", ".join(m.id for m in p.models)
        raise ValueError(
            f"Unknown model {model!r} for provider {provider!r}; "
            f"supported models: {allowed}."
        )

    if not _provider_available(config, p.id):
        env_var = "OPENAI_API_KEY" if p.id == "openai" else "GOOGLE_API_KEY"
        raise ValueError(
            f"Provider {p.id!r} is selected but {env_var} is not configured "
            "on the server."
        )

    return (p.id, model, _model_family(p.id, model))

File: desktop/api/src/model_catalog.rs
//! Catalog of provider/model pairs the API accepts at runtime. Ports
//! `micracode_core.model_catalog.list_catalog`, including the optional Ollama
//! probe and the default-selection logic.

use serde_json::{json, Value};

use crate::config::{Config, LlmProvider};

struct Model {
    id: &'static str,
    label: &'static str,
}

struct Provider {
    id: &'static str,
    label: &'static str,
    models: &'static [Model],
}

const PROVIDERS: &[Provider] = &[
    Provider {
        id: "openai",
        label: "OpenAI",
        models: &[
            Model { id: "gpt-5.4", label: "GPT-5.4" },
            Model { id: "gpt-5-mini", label: "GPT-5 Mini" },
            Model { id: "gpt-4.1", label: "GPT-4.1" },
        ],
    },
    Provider {
        id: "gemini",
        label: "Google Gemini",
        models: &[
            Model { id: "gemini-2.5-flash", label: "Gemini 2.5 Flash" },
            Model { id: "gemini-2.5-pro", label: "Gemini 2.5 Pro" },
            Model { id: "gemini-2.5-flash-lite", label: "Gemini 2.5 Flash Lite" },
        ],
    },
];

fn provider(pid: &str) -> Option<&'static Provider> {
    PROVIDERS.iter().find(|p| p.id == pid)
}

fn has_model(p: &Provider, model_id: &str) -> bool {
    p.models.iter().any(|m| m.id == model_id)
}

fn provider_available(config: &Config, pid: &str) -> bool {
    match pid {
        "openai" => !config.openai_api_key.is_empty(),
        "gemini" => !config.google_api_key.is_empty(),
        _ => false,
    }
}

async fn fetch_ollama_models(base_url: &str) -> Vec<String> {
    let client = match reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(3))
        .build()
    {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };
    let url = format!("{base_url}/api/tags");
    let resp = match client.get(&url).send().await {
        Ok(r) if r.status().is_success() => r,
        _ => return Vec::new(),
    };
    let body: Value = match resp.json().await {
        Ok(v) => v,
        Err(_) => return Vec::new(),
    };
    body.get("models")
        .and_then(Value::as_array)
        .map(|arr| {
            arr.iter()
                .filter_map(|m| m.get("name").and_then(Value::as_str).map(str::to_string))
                .collect()
        })
        .unwrap_or_default()
}

/// Serialise the registry for `GET /v1/models`.
pub async fn list_catalog(config: &Config) -> Value {
    let mut providers: Vec<Value> = PROVIDERS
        .iter()
        .map(|p| {
            json!({
                "id": p.id,
                "label": p.label,
                "available": provider_available(config, p.id),
                "models": p.models.iter().map(|m| json!({"id": m.id, "label": m.label})).collect::<Vec<_>>(),
            })
        })
        .collect();

    let ollama_models = fetch_ollama_models(&config.ollama_base_url).await;
    if !ollama_models.is_empty() {
        providers.push(json!({
            "id": "ollama",
            "label": "Ollama (local)",
            "available": true,
            "models": ollama_models.iter().map(|n| json!({"id": n, "label": n})).collect::<Vec<_>>(),
        }));
    }

    let (default_provider, default_model) = default_selection(config, &ollama_models);

    json!({
        "providers": providers,
        "default": { "provider": default_provider, "model": default_model },
    })
}

fn default_selection(config: &Config, ollama_models: &[String]) -> (String, String) {
    let env_model = config.active_model();

    match config.llm_provider {
        LlmProvider::Ollama => {
            if !env_model.is_empty() {
                return ("ollama".to_string(), env_model.to_string());
            }
            if let Some(first) = ollama_models.first() {
                return ("ollama".to_string(), first.clone());
            }
        }
        other => {
            if let Some(p) = provider(other.as_str()) {
                if !env_model.is_empty() && has_model(p, env_model) {
                    return (other.as_str().to_string(), env_model.to_string());
                }
            }
        }
    }

    for p in PROVIDERS {
        if provider_available(config, p.id) {
            if let Some(m) = p.models.first() {
                return (p.id.to_string(), m.id.to_string());
            }
        }
    }

    if let Some(first) = ollama_models.first() {
        return ("ollama".to_string(), first.clone());
    }

    let first = &PROVIDERS[0];
    (first.id.to_string(), first.models[0].id.to_string())
}
